networkx/algorithms/approximation/tests/test_traveling_salesman.py

```diff
@@ -402,3 +382,664 @@ def test_TSP_incomplete_graph_short_path():
+        G.add_edge(u, v)
+
+    gamma = tsp.spanning_tree_distribution(G, z_star)
+
+    assert {key: round(gamma[key], 4) for key in gamma} == solution_gamma
+
+
+def test_spanning_tree_sample():
+    """
+    Sample a single spanning tree from the distribution created in the last test
+    """
+    import networkx.algorithms.approximation.traveling_salesman as tsp
+    from math import exp
+    from random import Random
+
+    pytest.importorskip("numpy")
+    stats = pytest.importorskip("scipy.stats")
+
+    gamma = {
+        (0, 1): -0.6383,
+        (0, 2): -0.6827,
+        (0, 5): 0,
+        (1, 2): -1.0781,
+        (1, 4): 0,
+        (2, 3): 0,
+        (5, 3): -0.2820,
+        (5, 4): -0.3327,
+        (4, 3): -0.9927,
+    }
+
+    # The undirected support of gamma
+    G = nx.Graph()
+    for u, v in gamma:
+        if (u, v) in G.edges or (v, u) in G.edges:
+            continue
+        G.add_edge(u, v, lambda_key=exp(gamma[(u, v)]))
+
+    # Find the multiplicative weight for each tree.
+    total_weight = 0
+    tree_expected = {}
+    for t in nx.SpanningTreeIterator(G):
+        # Find the multiplicative weight of the spanning tree
+        weight = 1
+        for u, v, d in t.edges(data="lambda_key"):
+            weight *= d
+        tree_expected[t] = weight
+        total_weight += weight
+
+    # Assert that every tree has an entry in the expected distribution
+    assert len(tree_expected) == 75
+
+    # Set the sample size and then calculate the expected number of times we
+    # expect to see each tree. This test uses a near minimum sample size where
+    # the most unlikely tree has an expected frequency of 5.15.
+    # (Minimum required is 5)
+    #
+    # Here we also initialize the tree_actual dict so that we know the keys
+    # match between the two. We will later take advantage of the fact that since
+    # python 3.7 dict order is guaranteed so the expected and actual data will
+    # have the same order.
+    sample_size = 1200
+    tree_actual = {}
+    for t in tree_expected:
+        tree_expected[t] = (tree_expected[t] / total_weight) * sample_size
+        tree_actual[t] = 0
+
+    # Sample the spanning trees
+    #
+    # Assert that they are actually trees and record which of the 75 trees we
+    # have sampled.
+    #
+    # For repeatability, we want to take advantage of the decorators in NetworkX
+    # to randomly sample the same sample each time. However, if we pass in a
+    # constant seed to sample_spanning_tree we will get the same tree each time.
+    # Instead, we can create our own random number generator with a fixed seed
+    # and pass those into sample_spanning_tree.
+    rng = Random(37)
+    for _ in range(sample_size):
+        sampled_tree = tsp.sample_spanning_tree(G, "lambda_key", rng)
+        assert nx.is_tree(sampled_tree)
+
+        for t in tree_expected:
+            if nx.utils.edges_equal(t.edges, sampled_tree.edges):
+                tree_actual[t] += 1
+
```

Comment:
I'm not actually sure what the best way to handle this test in github actions. This test as written is a good check that the function does approximately what we want it to. But I'm worried about whether it should be in the unittests we run after every commit to a PR.  What is the best way to handle this?

We don't want to actually use a random test because that can lead to intermittent failures which are really hard to debug.  And if possible it would be good to test against the actual result provided for just one `sampled_tree` so that we can see if future changes have changed the way the method works.

I think we should at least  add a test to check that `tsp.sample_spanning_tree(G, "lambda_key", seed=42)` produces a fixed result (the result currently produced by this method).  And maybe we should comment out this test... Or name it something so that pytest doesn't automatically run it.

Reply:
Since this test passes in a `Random` instance for the trees, it is not actually a random fix. We are running the chi-squared test on a fixed sample, so it should be safe to run the test every commit. I cannot pass in an integer here as that would result in the same tree being sampled each time, just like in your propused test for sampling a signal spanning tree, so instead I use the `Random` instance to get a constant sequence of random numbers. Also, I like that new test idea.